receiver/awscontainerinsightreceiver/internal/gpu/decorator.go

```diff
@@ -0,0 +1,204 @@
+}
+
+func (gr *gpuMetric) GetTag(key string) string {
+	return gr.tags[key]
+}
+
+func (gr *gpuMetric) RemoveTag(key string) {
+	delete(gr.tags, key)
+}
+
+// GPU decorator acts as an interceptor of metrics before the scraper sends them to the next designated consumer
+type decorateConsumer struct {
+	containerOrchestrator string
+	nextConsumer          consumer.Metrics
+	k8sDecorator          Decorator
+	logger                *zap.Logger
+}
+
+func (dc *decorateConsumer) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{
+		MutatesData: true,
+	}
+}
+
+func (dc *decorateConsumer) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) error {
+	resourceTags := make(map[string]string)
+	rms := md.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		// get resource attributes
+		ras := rms.At(i).Resource().Attributes()
+		ras.Range(func(k string, v pcommon.Value) bool {
+			resourceTags[k] = v.AsString()
+			return true
+		})
+		ilms := rms.At(i).ScopeMetrics()
+		for j := 0; j < ilms.Len(); j++ {
+			ms := ilms.At(j).Metrics()
+			for k := 0; k < ms.Len(); k++ {
+				m := ms.At(k)
+				fields, tags := ci.ConvertToFieldsAndTags(m, dc.logger)
+				maps.Copy(tags, resourceTags)
+				rm := gpuMetric{
+					fields: fields,
+					tags:   tags,
+				}
+				if !rm.HasTag(ci.MetricType) {
+					// force type to be Container to decorate with container level labels
+					rm.AddTag(ci.MetricType, ci.TypeGpuContainer)
+				}
+				dc.decorateMetrics([]*gpuMetric{&rm})
+				dc.updateAttributes(m, rm)
+				if unit, ok := metricToUnit[m.Name()]; ok {
+					m.SetUnit(unit)
+				}
+			}
+		}
+	}
+	return dc.nextConsumer.ConsumeMetrics(ctx, md)
+}
+
+type Decorator interface {
+	Decorate(stores.CIMetric) stores.CIMetric
+	Shutdown() error
+}
+
+func (dc *decorateConsumer) decorateMetrics(metrics []*gpuMetric) []*gpuMetric {
+	var result []*gpuMetric
+	for _, m := range metrics {
+		// add tags for EKS
+		if dc.containerOrchestrator == ci.EKS {
+			out := dc.k8sDecorator.Decorate(m)
+			if out != nil {
+				result = append(result, out.(*gpuMetric))
+			}
+		}
+	}
+	return result
+}
+
+func (dc *decorateConsumer) updateAttributes(m pmetric.Metric, gm gpuMetric) {
+	if len(gm.tags) < 1 {
+		return
+	}
+	var dps pmetric.NumberDataPointSlice
+	switch m.Type() {
+	case pmetric.MetricTypeGauge:
+		dps = m.Gauge().DataPoints()
+	case pmetric.MetricTypeSum:
+		dps = m.Sum().DataPoints()
+	default:
+		dc.logger.Warn("Unsupported metric type", zap.String("metric", m.Name()), zap.String("type", m.Type().String()))
+	}
+
+	if dps.Len() < 1 {
+		return
+	}
+	attrs := dps.At(0).Attributes()
+	for tk, tv := range gm.tags {
+		// type gets set with metrictransformer while duplicating metrics at different resource levels
+		if tk == ci.MetricType {
+			continue
+		}
+		attrs.PutStr(tk, tv)
+	}
+}
+
+func (dc *decorateConsumer) Shutdown() error {
+	var errs error
+
+	if dc.k8sDecorator != nil {
+		errs = errors.Join(errs, dc.k8sDecorator.Shutdown())
```

Comment:
nit: Can just return `dc.k8sDecorator.Shutdown()`. There's no other error to join.

Reply:
true, updated